app/jobs/application_job.rb

```diff
@@ -21,16 +20,69 @@ class ApplicationJob < ActiveJob::Base
     throw(:abort) unless job.class.before_enqueue_lock(*job.arguments)
   end
 
-  # Overriding so that changes in ActiveModel object don't result in new lock (which they do when just calling to_s).
-  def self.lock(*args)
-    new_args = args.map { |arg| arg.class.method_defined?(:to_global_id) ? arg.to_global_id.to_s : arg }
-    "lock:#{self.class.name}-#{new_args}"
+  around_perform do |job, block|
+    block.call
+  ensure
+    # Always clear the lock when we're done, even if there was an error.
+    job.class.clear_lock(*job.arguments)
   end
 
   # Override in subclass to tune per queue.
-  # def lock_timeout
-  #   3600
-  # end
+  # @return [Integer] lock_timeout in seconds, after which lock is no longer valid,
+  # regardless of whether the job that created the lock is still in queue or working.
+  def self.lock_timeout
+    # 1 hour is a reasonable default, because most jobs in this app will
+    # cause little if re-executed too aggressively.
+    3600
+  end
+
+  # @return [String] the key for locking this job/payload combination, e.g. 'lock:MySpecificJob-bt821jk7040;1'
+  def self.queue_lock_key(*args)
+    # Changes in ActiveModel object don't result in new lock (which they do when just calling to_s).
+    queue_lock_args = args.map { |arg| arg.class.method_defined?(:to_global_id) ? arg.to_global_id.to_s : arg }
+    "lock:#{name}-#{queue_lock_args.join(';')}"
+  end
+
+  def self.before_enqueue_lock(*args) # rubocop:disable Metrics/AbcSize
```

Comment:
this method contains the trickiest and most important part of the resque-lock logic.  i tried to keep it basically the same as the original in terms of its use of redis and the conditions it checks, but i did rename some variables and add some logging for clarity.